DiskCache: `Get` behavior for empty directories is inconsistent across filesystems

### What happened?

The current implementation of `DiskCache.Get` uses `os.FileInfo.Size()` to determine whether a cached entry should be treated as “not existing”:

```go
if fi.Size() == 0 {
    return p, os.ErrNotExist
}
```
However, this behavior is inconsistent across filesystems and environments. Especially when running in a container with an overlayfs:

```sh
# running a docker env
$ docker run --rm debian:trixie bash -c "mkdir /tmp/test && stat -c '%n size=%s type=%F' /tmp/test/ && df -T /tmp/test/"
/tmp/test/ size=0 type=directory
Filesystem     Type    1K-blocks      Used Available Use% Mounted on
overlay        overlay 975097856 723207396 227207548  77% /
```

--> the directory size of `/tmp/test` in my debian container is **0 bytes**.

```sh
# running on host
$ mkdir /tmp/test && stat -c '%n size=%s type=%F' /tmp/test/ && df -T /tmp/test/
/tmp/test/ size=40 type=directory
Filesystem     Type  1K-blocks  Used Available Use% Mounted on
tmpfs          tmpfs  16386920   368  16386552   1% /tmp
```

--> the directory size of `/tmp/test` on my fedora system is **40 bytes**. 

This then also breaks the unit test `TestDiskCache_PutAndGet/GetDirectory` when building Helm in a container, as the `fi.Size() == 0` check fails and continues to `if fi.IsDir() { ... }`, which then returns the wrong error.

```sh
$ docker run --rm --volume .:/helm -w /helm -it golang:1.26-alpine go test -test.fullpath=true -timeout 30s -run ^TestDiskCache_PutAndGet$ helm.sh/helm/v4/pkg/downloader
...
--- FAIL: TestDiskCache_PutAndGet (0.00s)
    --- FAIL: TestDiskCache_PutAndGet/GetDirectory (0.00s)
        /helm/pkg/downloader/cache_test.go:112: 
                Error Trace:    /helm/pkg/downloader/cache_test.go:112
                Error:          Error message not equal:
                                expected: "is a directory"
                                actual  : "file does not exist"
                Test:           TestDiskCache_PutAndGet/GetDirectory
FAIL
FAIL    helm.sh/helm/v4/pkg/downloader  0.022s
FAIL
```

### What did you expect to happen?

That building Helm v4* works in a containerized environment.

### How can we reproduce it (as minimally and precisely as possible)?

To run the unit test in an Alpine container, execute:
```sh
git clone --depth 1 --branch v4.1.4 https://github.com/helm/helm.git && cd helm/
docker run --rm --volume .:/helm -w /helm -it golang:1.26-alpine go test -test.fullpath=true -timeout 30s -run ^TestDiskCache_PutAndGet$ helm.sh/helm/v4/pkg/downloader
```

### Helm version

Working with Helm v4.1.4.

<details>

```console
~/dev/helm$ git status
HEAD detached at v4.1.4
```
</details>


### Kubernetes version

None

### Potential fix

To fix the unit test error, we could just change the ordering of the checks, instead of checking `fi.Size() == 0` first, check `fi.IsDir()` first:
```go
// Get returns a reader for the given key.
func (c *DiskCache) Get(key [sha256.Size]byte, cacheType string) (string, error) {
	p := c.fileName(key, cacheType)
	fi, err := os.Stat(p)
	if err != nil {
		return "", err
	}
	// directories should never happen unless something outside helm is operating
	// on this content.
	if fi.IsDir() {
		return p, errors.New("is a directory")
	}
	// Empty files treated as not exist because there is no content.
	if fi.Size() == 0 {
		return p, os.ErrNotExist
	}
	return p, nil
}
```

A different approach would be to actually check the content of the file.

If you want, I can create a PR for your desired change.

--- a/pkg/downloader/cache.go
+++ b/pkg/downloader/cache.go
@@ -59,14 +59,16 @@ func (c *DiskCache) Get(key [sha256.Size]byte, cacheType string) (string, error)
 	if err != nil {
 		return "", err
 	}
-	// Empty files treated as not exist because there is no content.
-	if fi.Size() == 0 {
-		return p, os.ErrNotExist
-	}
 	// directories should never happen unless something outside helm is operating
 	// on this content.
 	if fi.IsDir() {
 		return p, errors.New("is a directory")
+	}
+	// Empty files treated as not exist because there is no content.
+	// IsDir must be checked first: some filesystems (e.g. overlay) report
+	// directory size as 0.
+	if fi.Size() == 0 {
+		return p, os.ErrNotExist
 	}
 	return p, nil
 }

--- a/pkg/downloader/cache.go
+++ b/pkg/downloader/cache.go
@@ -64,8 +64,8 @@ func (c *DiskCache) Get(key [sha256.Size]byte, cacheType string) (string, error)
 	if fi.IsDir() {
 		return p, errors.New("is a directory")
 	}
-	// Empty files treated as not exist because there is no content.
-	// IsDir must be checked first: some filesystems (e.g. overlay) report
+	// Empty files are treated as non-existent because there is no content.
+	// IsDir must be checked first: some filesystems (e.g. overlayfs) report
 	// directory size as 0.
 	if fi.Size() == 0 {
 		return p, os.ErrNotExist

